scio-extra/src/main/scala/com/spotify/scio/extra/voyager/package.scala

```diff
@@ -0,0 +1,291 @@
+     * @param spaceType
+     *   The measurement for computing distance between entities. One of Euclidean, Cosine or Dot
+     *   (inner product).
+     * @param storageDataType
+     *   The storage data type of the vectors at rest. One of Float8, Float32 or E4M3.
+     * @param dim
+     *   Number of dimensions in vectors.
+     * @param ef
+     *   The size of the dynamic list of neighbors used during construction time. This parameter
+     *   controls query time/accuracy tradeoff. More information can be found in the hnswlib
+     *   documentation https://github.com/nmslib/hnswlib.
+     * @param m
+     *   The number of outgoing connections in the graph.
+     * @return
+     *   A [[VoyagerUri]] representing where the index was written to.
+     */
+    @experimental
+    def asVoyager(
+      path: String,
+      spaceType: SpaceType,
+      storageDataType: StorageDataType,
+      dim: Int,
+      ef: Long,
+      m: Long
+    ): SCollection[VoyagerUri] = {
+      val uri: VoyagerUri = VoyagerUri(path, self.context.options)
+      require(!uri.exists, s"Voyager URI ${uri.path} already exists")
+      logger.info(s"Voyager URI :${uri.path}")
+      self.transform { in =>
+        {
+          in.groupBy(_ => ())
+            .map { case (_, xs) =>
+              val voyagerWriter: VoyagerWriter =
+                new VoyagerWriter(spaceType, storageDataType, dim, ef, m)
+
+              voyagerWriter.write(xs)
+              uri.saveAndClose(voyagerWriter)
+              uri
+            }
+        }
+      }
+    }
+
+    /**
+     * Write the key-value pairs of this SCollection as a Voyager index to a temporary location and
+     * building the index using the parameters specified.
+     *
+     * @param distanceMeasure
+     *   The measurement for computing distance between entities. One of Euclidean, Cosine or Dot
+     *   (inner product).
+     * @param storageDataType
+     *   The Storage type of the vectors at rest. One of Float8, Float32 or E4M3.
+     * @param dim
+     *   Number of dimensions in vectors.
+     * @param ef
+     *   The size of the dynamic list of neighbors used during construction time. This parameter
+     *   controls query time/accuracy tradeoff. More information can be found in the hnswlib
+     *   documentation https://github.com/nmslib/hnswlib.
+     * @param m
+     *   The number of outgoing connections in the graph.
+     * @return
+     *   A [[VoyagerUri]] representing where the index was written to.
+     */
+    @experimental
+    def asVoyager(
+      distanceMeasure: SpaceType,
+      storageDataType: StorageDataType,
+      dim: Int,
+      ef: Long = 200L,
+      m: Long = 16L
+    ): SCollection[VoyagerUri] = {
+      val uuid: UUID = UUID.randomUUID()
+      val tempLocation: String = self.context.options.getTempLocation
+      require(tempLocation != null, s"--tempLocation arg is required")
+      val path = s"$tempLocation/voyager-build-$uuid"
```

Comment:
I think it would be a little safer to use `Paths#resolve` here, in case tempLocation is passed with a trailing slash -- i.e. `Paths.get(tempLocation).resolve(s"voyager-build-$uuid")`

Reply:
I can try that, I was getting odd issues in prod with `resolve` cutting out a part of the `gs://filename` and turning it into `gs:/filename`.

Reply:
ah yes so this is what happens `Exception in thread "main" java.lang.IllegalArgumentException: Invalid GCS URI: gs:/bucket-name/temp/voyager-build-b9818c3b-2700-4a66-b281-e6ab863d7547/index.hnsw`. Is there a way to fix that or make the `Paths` object respect that the gcs uri?

Reply:
I added in a `.stripSuffix("/")` which ideally should handle that case?